### DISCUSSION TITLE
Fillet functions and understanding OverOffsets

### DISCUSSION THREAD

**Discussion by eflehder1** [leap71/PicoGK#44] [Coding for (Computational) Engineers]
Hi @josefinelissner @LinKayser, with a feature like the one shown below, would the best way to ensure a smooth transition from one part to the next be to use multiple shapekernel frames with a spine, or is there a 'fillet' feature that can automatically create this transition between two parts merged via BoolAdd?

![image](https://github.com/user-attachments/assets/b1e58d2a-1978-4297-8f92-ab200cb2fe1d)

**Comment by josefinelissner** [comment]
Hi, @eflehder1 

Since we are dealing with voxelfields and not with individual surface patches, it is not easy to obtain a single edge fillet. However, we have a bunch of functions to smooth out sharp-edged corners. Since they are always applied globally on the entire voxelfield, you need to consider in your algorithm when a smoothening step is suitable. Usually, I keep "fine-detail" voxelfields separate from "low-detail" voxelfields (like bounding shapes and chunky parts) until the very end of the algorithm. This way I can smooth the coarse bits multiple times without destroying fine details.

I recommend to check out the [OverOffsetShowCase.cs](https://github.com/leap71/LEAP71_ShapeKernel/blob/main/Examples/Ex_OverOffsetShowCase.cs) in the ShapeKernel examples folder.

It helps you to visualize how you can use the OverOffset function to achieve rounded corners. The basic Offset function is not reversible as it "destroys" details below the offset value threshold. This means that if you offset outwards by distance x and then you offset inwards by the same amount, you will not have the same shape as before. This is what we call OverOffset. I am using this function to selectively "destroy" convex or concave details on my voxelfield. If you offset outwards by 3 and then back to 0 you have roundings of R=3mm on all concave parts of the geometry. If you offset inwards by 3 (-3mm) and then back to 0 you have roundings of R=3mm on all convex parts. If you combine these two OverOffsets you "destroy" convex and concave details alike and you arrive at our Smoothen function.

<img width="1011" alt="image" src="https://github.com/user-attachments/assets/74f3a5c8-52dd-4b4b-b356-ba811abf21c4">

This is a very powerful toolset. It needs some practice to apply it correctly in your code. I hope this helps!

**Reply by eflehder1** [reply]
This is really helpful thank you very much @josefinelissner! I will explore the use of these functions, it will actually work quite well for this part in that example because the only concave edge is the one I want to create the fillet for.

I worked a bit more on that specific problem I was dealing with and came up with a strategy to more or less create the geometry I was looking for, making use of one of your showcase examples. My approach: 1) Create a line that defines the skeleton, 2) Derive frames from this line, 3) Create a base cylinder using these frames, 4) Set the radius with a surface modulation which consists of a radial function for the first 60% of the length. The result is as shown below:

![image](https://github.com/user-attachments/assets/ad027205-08a7-4fd2-8189-88ee04dae2e4)
![image](https://github.com/user-attachments/assets/9520fd73-7011-46cc-9f23-213ef281852a)

I wanted to ask some questions about this workflow: a) I intially tried using a ControlPointSpline as shown in one of your examples, however, the frames would then change orientation and I needed the first frame to be normal to that initial plane. Is it possible to control the orientation of each frame independently? b) my final geometry is sort of staggered, I have just tried to smoothen it using the overoffset function you recommended and it does work but as you mentioned, the change is applied globally, how would one go about ensuring the smoothening is only applied to selected convex edges?

![image](https://github.com/user-attachments/assets/38a4254d-ec60-4483-b02c-b3fad0f814a1)

**Reply by soigniert** [reply]
Hi,
OpenVDB provides functionality to apply voxel offsets/smoothing [locally using a "Mask" natively](https://www.openvdb.org/documentation/doxygen/classopenvdb_1_1v12__0_1_1tools_1_1LevelSetFilter.html#details). With a mask one can selectively apply the operation position and strength. See an example here:
<img width="1251" height="1068" alt="ViewCapture20250804_031553" src="https://github.com/user-attachments/assets/6f2985ea-5342-4b43-b795-141b20f11a2e" />

Actually I would love to see this functionality implemented into PicoGK. With the minMas() maxMask() functionionality I can image to convert lattice elements with different max values to masks in order to even create a modulated/changing fillet function or to use implicit elements to prevent certain regions from smoothing. Maybe at some point leap71 finds the functionality worth to implement :)
